### DISCUSSION TITLE
Is it possible to override behaviour for IME_ACTION_PREVIOUS in ReactTextInputManager?

### DISCUSSION THREAD

**Discussion by ant100** [react-native-tvos/react-native-tvos#641] [Q&A] [answered]
Taking a look at [this line](https://github.com/react-native-tvos/react-native-tvos/blob/tvos-v0.72.4/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputManager.java#L1117) it seems like all of actions in IME_MASK_ACTION trigger the submit event. I'd like to override this behaviour to make the IME_ACTION_PREVIOUS action work as a "back" button and dismisses the keyboard. 

Does anybody know if this is possible? I managed to get my desired behaviour by directly modifying the file inside node_modules, but would like a way to override it from my project. 

I'm also wondering why is this not the default behaviour. Is there a use case for when the IME_ACTION_PREVIOUS action should work the same as IME_ACTION_DONE? 

Working on react-native-tvos@0.72.6-0

**Comment by douglowder** [comment, marked as answer]
@ant100 this section of the text input native Android code is unchanged from React Native core, and appears to have been there for many years. I would not touch that code.

If you are able to make things work the way you want by modifying `BackHandler.android.js` then that is probably the best approach, and you can use `patch-package` or similar module to make your patch automatically whenever you do a yarn install.

**Reply by ant100** [reply]
@douglowder  Thanks so much for your answer Doug!! I'm taking a look at `BackHandler.android.js`, the problem seems to be that BackHandler doesn't work when the keyboard is invoked, I'm suspecting due to IME service taking over. I'll try to see if there's a way around this, and come back with a solution if possible.

Also, if you happen to have any idea on how I might achieve this I'd really appreciate it. I'm navigating in the dark here. 

Thanks again!

Edit: After a bit more research, seems like when we have the keyboard open, the Input Media Editor captures the press event before it gets to the BackHandler. A solution would be to be able to override the `onKeyPreIme` method, but again not sure if this is possible. Will keep looking.

**Reply by ant100** [reply]
Update 2: 
I found a way to override `onKeyPreIme` method from [ReactEditText.java](https://github.com/react-native-tvos/react-native-tvos/blob/a49cae2940c9e6eb90728b6132e1c75d11b8843d/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditText.java),  unfortunately this approach doesn't work because it only listens to the hardware key press but not to the key event from the soft keyboard. Meaning: I can catch the event if I press the "previous" button in my remote, but not if I select the on-screen "previous" button from the soft keyboard.

So far, the only way I've found to capture both events (the previous button from the remote and previous button from keyboard) is to modify the `onEditorAction` override in the `ReactTextInputManager`. Even this approach is not working perfectly yet, because keyboard is dismissed but it can't regain focus. 

I'm concerned about modifying this part of the code, but haven't found a different place where we can override the IME actions.
